README.md
# AskOverflow

<p align="center"><img src="media/askoverflow_icon.png" /></p>

<p align="center">The most powerful developer website, inside your favorite editor.</p>

<p align="center"><a href="https://www.buymeacoffee.com/fabiosabbion"><img alt="buy me a coffe" src="media/buymeacoffe-button.png" style="border-radius: 16px;"/></a></p>

## Features

### Ask questions

Ask questions to Stackoverflow without leaving your editor.
Append your code to the body of the question using the "Insert code from current selection" command.

![ask-feature-gif](https://user-images.githubusercontent.com/44337776/137006174-7ad9a8e7-7db4-44cc-9f11-2cdf66180dd6.gif)

### Search questions

Search for questions on Stackoverflow.

![search-feature-gif](https://user-images.githubusercontent.com/44337776/137006588-5f19bdbb-c064-4dad-b9be-7500d65bc769.gif)

### Monitor your questions

Keep an eye on your posted questions: whether they got answered, how many times they got viewed, and more.


CHANGELOG.md
# Change Log

## 1.1.1 (latest)

- fixed bug in My Questions Page

## 1.1.0

- added "My Questions" page
- fixed bugs

## 1.0.1

- fixed text encoding

## 1.0.0

- added search feature

## 0.1.0

- initial release


api/index.js
const clientId = 20981;
const scope = "write_access";
const baseUrl = "https://xeyuug.deta.dev";

const clientSecret = require("./secret");
const express = require('express');
const request = require('request');
const app = express();

app.get("/debug", (req, res) => {
    res.end("DEBUG PAGE WORKING");
});

app.get("/redirect", (req, res) => {
    let code = req.query.code;
    if (code) {
        const options = {
            url: "https://stackoverflow.com/oauth/access_token/json?client_id=" + clientId + "&client_secret=" + clientSecret + "&code=" + code + "&redirect_uri=" + encodeURI(baseUrl) + "/redirect",
            method: 'POST'
        };

        request(options, (err, _, body) => {
            if (err) {
                res.redirect("http://localhost:56789/auth/0");
            } else {
                let data = JSON.parse(body);
                const token = data.access_token;
                res.redirect("http://localhost:56789/auth/" + token);
            }
        });


    } else {
        res.redirect("http://localhost:56789/auth/0");
    }

});

app.get("/auth", (req, res) => {
    res.redirect("https://stackoverflow.com/oauth?client_id=" + clientId + "&scope=" + scope + "&redirect_uri=" + encodeURI(baseUrl) + "/redirect");

});

module.exports = app;

src/extension.ts
import * as vscode from 'vscode';
import { QuestionManager } from './QuestionManager';
import { SidebarProvider } from './SidebarProvider';
import { TokenManager } from './TokenManager';

export function activate(context: vscode.ExtensionContext) {

	TokenManager.initAsync(context);
	QuestionManager.initAsync(context);

	const sidebarProvider = new SidebarProvider(context.extensionUri);
	context.subscriptions.push(
		vscode.window.registerWebviewViewProvider(
			"askoverflow-sidebar",
			sidebarProvider
		)
	);

	context.subscriptions.push(vscode.commands.registerCommand('askoverflow.insertCode', () => {
		let editor = vscode.window.activeTextEditor;

		if (editor === undefined) {
			vscode.window.showErrorMessage('No active text editor');
			return;
		}
		let text = editor.document.getText(editor.selection);
		if (text === "") {
			vscode.window.showInformationMessage('Selection is empty');
			return;
		}
		sidebarProvider._view?.webview.postMessage({ type: "selectedCode", value: text });
	}));

	// context.subscriptions.push(vscode.commands.registerCommand('askoverflow.refresh', async () => {
	// 	await vscode.commands.executeCommand("workbench.action.closeSidebar");
	// 	await vscode.commands.executeCommand("workbench.view.extension.askoverflow-sidebar-view");
	// 	setTimeout(() => { vscode.commands.executeCommand("workbench.action.webview.openDeveloperTools"); }, 100);

	// }));

}

// this method is called when your extension is deactivated
export function deactivate() { }


src/HelloWorldPanel.ts
import * as vscode from "vscode";

export class HelloWorldPanel {
    /**
     * Track the currently panel. Only allow a single panel to exist at a time.
     */
    public static currentPanel: HelloWorldPanel | undefined;

    public static readonly viewType = "hello-world";

    private readonly _panel: vscode.WebviewPanel;
    private readonly _extensionUri: vscode.Uri;
    private _disposables: vscode.Disposable[] = [];

    public static createOrShow(extensionUri: vscode.Uri) {
        const column = vscode.window.activeTextEditor
            ? vscode.window.activeTextEditor.viewColumn
            : undefined;

        // If we already have a panel, show it.
        if (HelloWorldPanel.currentPanel) {
            HelloWorldPanel.currentPanel._panel.reveal(column);
            HelloWorldPanel.currentPanel._update();
            return;
        }

        // Otherwise, create a new panel.
        const panel = vscode.window.createWebviewPanel(
            HelloWorldPanel.viewType,
            "HelloWorld",
            column || vscode.ViewColumn.One,
            {
                // Enable javascript in the webview
                enableScripts: true,

                // And restrict the webview to only loading content from our extension's `media` directory.
                localResourceRoots: [
                    vscode.Uri.joinPath(extensionUri, "media"),
                    vscode.Uri.joinPath(extensionUri, "out/compiled"),
                ],
            }
        );

        HelloWorldPanel.currentPanel = new HelloWorldPanel(panel, extensionUri);
    }

    public static kill() {
        HelloWorldPanel.currentPanel?.dispose();
        HelloWorldPanel.currentPanel = undefined;
    }

    public static revive(panel: vscode.WebviewPanel, extensionUri: vscode.Uri) {
        HelloWorldPanel.currentPanel = new HelloWorldPanel(panel, extensionUri);
    }

    private constructor(panel: vscode.WebviewPanel, extensionUri: vscode.Uri) {
        this._panel = panel;
        this._extensionUri = extensionUri;

        // Set the webview's initial html content
        this._update();

        // Listen for when the panel is disposed
        // This happens when the user closes the panel or when the panel is closed programatically
        this._panel.onDidDispose(() => this.dispose(), null, this._disposables);

        // // Handle messages from the webview
        // this._panel.webview.onDidReceiveMessage(
        //   (message) => {
        //     switch (message.command) {
        //       case "alert":
        //         vscode.window.showErrorMessage(message.text);
        //         return;
        //     }
        //   },
        //   null,
        //   this._disposables
        // );
    }

    public dispose() {
        HelloWorldPanel.currentPanel = undefined;

        // Clean up our resources
        this._panel.dispose();

        while (this._disposables.length) {
            const x = this._disposables.pop();
            if (x) {
                x.dispose();
            }
        }
    }

    private async _update() {
        const webview = this._panel.webview;

        this._panel.webview.html = this._getHtmlForWebview(webview);
        webview.onDidReceiveMessage(async (data) => {
            switch (data.type) {
                case "onInfo": {
                    if (!data.value) {
                        return;
                    }
                    vscode.window.showInformationMessage(data.value);
                    break;
                }
                case "onError": {
                    if (!data.value) {
                        return;
                    }
                    vscode.window.showErrorMessage(data.value);
                    break;
                }
                //case "tokens": {
                //   await Util.globalState.update(accessTokenKey, data.accessToken);
                //   await Util.globalState.update(refreshTokenKey, data.refreshToken);
                //   break;
                //}
            }
        });
    }

    private _getHtmlForWebview(webview: vscode.Webview) {
        // // And the uri we use to load this script in the webview
        const scriptUri = webview.asWebviewUri(
            vscode.Uri.joinPath(this._extensionUri, "out", "compiled/HelloWorld.js")
        );

        // Uri to load styles into webview
        const stylesResetUri = webview.asWebviewUri(vscode.Uri.joinPath(
            this._extensionUri,
            "media",
            "reset.css"
        ));
        const stylesMainUri = webview.asWebviewUri(vscode.Uri.joinPath(
            this._extensionUri,
            "media",
            "vscode.css"
        ));

        //const cssUri = webview.asWebviewUri(
        //    vscode.Uri.joinPath(this._extensionUri, "out", "compiled/swiper.css")
        //);

        // Use a nonce to only allow specific scripts to be run
        const nonce = getNonce();

        return `<!DOCTYPE html>
			<html lang="en">
			<head>
				<meta charset="UTF-8">
				<!--
					Use a content security policy to only allow loading images from https or from our extension directory,
					and only allow scripts that have a specific nonce.
        -->
        <meta http-equiv="Content-Security-Policy" content="img-src https: data:; style-src 'unsafe-inline' ${webview.cspSource
            }; script-src 'nonce-${nonce}';">
				<meta name="viewport" content="width=device-width, initial-scale=1.0">
				<link href="${stylesResetUri}" rel="stylesheet">
				<link href="${stylesMainUri}" rel="stylesheet">
       
        <script nonce="${nonce}">
            
        </script>
		</head>
        <body>
            <script nonce="${nonce}" src="${scriptUri}"></script>
		</body>
        
		</html>`;
    }
}

function getNonce() {
    let text = '';
    const possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
    for (let i = 0; i < 32; i++) {
        text += possible.charAt(Math.floor(Math.random() * possible.length));
    }
    return text;
}

src/SidebarProvider.ts
import { join } from "path";
import polka = require("polka");
import * as vscode from "vscode";
import { TokenManager } from './TokenManager';
import { QuestionManager } from './QuestionManager';
const apiBaseUrl = "https://xeyuug.deta.dev";

export class SidebarProvider implements vscode.WebviewViewProvider {
    _view?: vscode.WebviewView;
    _doc?: vscode.TextDocument;



    constructor(private readonly _extensionUri: vscode.Uri) { }

    public resolveWebviewView(webviewView: vscode.WebviewView) {
        this._view = webviewView;
        webviewView.webview.options = {
            // Allow scripts in the webview
            enableScripts: true,

            localResourceRoots: [this._extensionUri],
        };

        webviewView.webview.html = this._getHtmlForWebview(webviewView.webview);

        const app = polka();

        app.get("/auth/:token", (req, res) => {
            app.server?.close();
            let token = req.params.token;
            if (!token || token === "0") {
                res.end("auth failed");
                return;
            }
            this._view?.webview.postMessage({ type: "token", value: token });
            TokenManager.setToken(token);
            console.log("Token: " + token);
            res.end("auth success, you can close this window now");

        });

        webviewView.webview.onDidReceiveMessage(async (data) => {
            switch (data.type) {
                case "onAuth": {
                    // data.value is True for cached results, False otherwise
                    if (data.value) {
                        const cachedToken = TokenManager.getToken();
                        if (cachedToken) {
                            this._view?.webview.postMessage({ type: "token", value: cachedToken });
                            return;
                        }
                    }
                    app.listen(56789, () => {
                        console.log("running polka on localhost:56789");
                        vscode.commands.executeCommand("vscode.open", vscode.Uri.parse(apiBaseUrl + "/auth"));
                    });
                    break;
                }
                case "onNewQuestion": {
                    let id = data.value;
                    QuestionManager.addQuestion(id);
                    break;
                }
                case "onGetQuestions": {
                    let questionIds = QuestionManager.getQuestions();
                    this._view?.webview.postMessage({ type: "questions", value: questionIds });
                    break;
                }
                case "onInfo": {
                    if (!data.value) {
                        return;
                    }
                    vscode.window.showInformationMessage(data.value);
                    break;
                }
                case "onError": {
                    if (!data.value) {
                        return;
                    }
                    vscode.window.showErrorMessage(data.value);
                    break;
                }
            }
        });



    }

    public revive(panel: vscode.WebviewView) {
        this._view = panel;
    }

    private _getHtmlForWebview(webview: vscode.Webview) {
        console.log(this._extensionUri.path);
        const styleResetUri = webview.asWebviewUri(
            vscode.Uri.file(join(this._extensionUri.path, "media", "reset.css"))
        );
        const styleVSCodeUri = webview.asWebviewUri(
            vscode.Uri.file(join(this._extensionUri.path, "media", "vscode.css"))
        );

        const scriptUri = webview.asWebviewUri(
            vscode.Uri.file(join(this._extensionUri.path, "out", "compiled/sidebar.js"))
        );
        const styleMainUri = webview.asWebviewUri(
            vscode.Uri.file(join(this._extensionUri.path, "out", "compiled/sidebar.css"))
        );

        // Use a nonce to only allow a specific script to be run.
        const nonce = getNonce();

        return `<!DOCTYPE html>
			<html lang="en">
			<head>
				<meta charset="UTF-8">
				<!--
					Use a content security policy to only allow loading images from https or from our extension directory,
					and only allow scripts that have a specific nonce.
        -->
        <meta http-equiv="Content-Security-Policy" content="img-src https: data:; style-src 'unsafe-inline' ${webview.cspSource
            }; script-src 'nonce-${nonce}';">
				<meta name="viewport" content="width=device-width, initial-scale=1.0">
				<link href="${styleResetUri}" rel="stylesheet">
				<link href="${styleVSCodeUri}" rel="stylesheet">
                <link href="${styleMainUri}" rel="stylesheet">
                <script nonce="${nonce}">
                    const tsvscode = acquireVsCodeApi();
                </script>

			</head>
            <body>
				<script nonce="${nonce}" src="${scriptUri}"></script>
			</body>
			</html>`;
    }
}

function getNonce() {
    let text = '';
    const possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
    for (let i = 0; i < 32; i++) {
        text += possible.charAt(Math.floor(Math.random() * possible.length));
    }
    return text;
}